frontend/src/liboperator/operator/devices.go

```diff
@@ -103,22 +117,33 @@ func (d *Device) ToClient(id int, msg interface{}) error {
 	return client.Send(msg)
 }
 
+// preDevice holds information about a device between pre-registration and registration.
+type preDevice struct {
+	Desc  DeviceDesc
+	RegCh chan bool
+}
+
 // Keeps track of the registered devices.
 type DevicePool struct {
+	// preDevicesMtx must always be locked before devicesMtx
```

Comment:
rather than having two mutexes and two maps in the DevicePool data structure, would having a single mutex/map pair to handle pre registration and devices work here? Suggesting this as it would be easier to keep track of one mutex.


```
type preRegistration struct {
	desc  DeviceDesc
        done chan bool
}

type poolEntry {
  preReg *preRegistration
  device *Device
}

type DevicePool {
  mtx sync.Mutex
  entries  map[string]*poolEntry
}
```

Reply:
I thought about this and ended up preferring the two mutexes approach by a very slight margin. These are the reasons:
* The `done` channel makes no sense after registration, that fact is easily conveyed by it going out of scope
* The DeviceDescriptor is an integral part of the Device, the fact that it was provided at pre-registration is just a detail
* [minor] Listing registered devices with two maps doesn't require filtering
* [minor] No need to have a status field (or equivalent) to know if a device is pre-registered or fully registered

An alternative would be to use the same mutex for both maps. The downside of that is that interactions with registered devices (signaling) can't happen concurrently with pre-registration operations.

I'm ok with using the same mutex for both structs, but I strongly prefer having two structs over just one. WDYT?